jans-cedarling/cedarling/src/jwt/decoding_strategy/key_service.rs

```diff
@@ -0,0 +1,137 @@
+        for endpoint in openid_conf_endpoints {
+            // TODO: use a reusable client when making requests
+            let conf_src: OpenIdConfigSource = get(endpoint)
+                .map_err(Error::HttpError)?
+                .error_for_status()
+                .map_err(Error::HttpError)?
+                .json()
+                .map_err(Error::RequestDeserializationError)?;
+            let (issuer, conf) = OpenIdConfig::from_source(conf_src);
+            idp_configs.insert(issuer, conf);
+        }
+
+        /// retrieves a decoding key based on the provided key ID (`kid`).
+        ///
+        /// this method first attempts to retrieve the key from the local key store. if the key
+        /// is not found, it will refresh the JWKS and try again. if the key is still not found,
+        /// an error of type `KeyNotFound` is returned.
+        for (iss, conf) in &mut idp_configs {
+            let jwks: JwkSet = get(&*conf.jwks_uri)
+                .map_err(Error::HttpError)?
+                .error_for_status()
+                .map_err(Error::HttpError)?
+                .json()
+                .map_err(Error::RequestDeserializationError)?;
+            let mut decoding_keys = conf.decoding_keys.write().unwrap();
+            for jwk in jwks.keys {
+                let decoding_key = DecodingKey::from_jwk(&jwk).map_err(Error::KeyParsingError)?;
+                let key_id = jwk.common.key_id.ok_or(Error::MissingKeyId)?;
+                decoding_keys.insert(key_id.into(), Arc::new(decoding_key));
+            }
+        }
+
+        Ok(Self { idp_configs })
+    }
+
+    /// retrieves a decoding key based on the provided key ID (`kid`).
+    ///
+    /// this method first attempts to retrieve the key from the local key store. if the key
+    /// is not found, it will refresh the JWKS and try again. if the key is still not found,
+    /// an error of type `KeyNotFound` is returned.
+    pub fn get_key(&self, kid: &str) -> Result<Arc<DecodingKey>, Error> {
+        for (iss, config) in &self.idp_configs {
+            // first try to get the key from the local keystore
+            if let Some(key) = self.get_key_from_iss(&iss, &kid) {
+                return Ok(key.clone());
+            } else {
+                println!("could not find {}, updating jwks", kid);
+                // if the key is not found in the local keystore, update
+                // the local keystore and try again
+                self.update_jwks(iss);
+                if let Some(key) = self.get_key_from_iss(&iss, &kid) {
+                    return Ok(key.clone());
+                }
+            }
+        }
+
+        Err(Error::KeyNotFound(kid.into()))
+    }
+
+    /// helper function to retrieve a key for a specific issuer (`iss`).
+    fn get_key_from_iss(&self, iss: &str, kid: &str) -> Option<Arc<DecodingKey>> {
+        if let Some(idp) = self.idp_configs.get(iss) {
+            let decoding_keys = idp.decoding_keys.read().unwrap();
```

Comment:
maybe you can manage better the error here

Reply:
added a custom error. see: [460bc0b](https://github.com/JanssenProject/jans/pull/9855/commits/460bc0bf241f11a896fa6c8482bd72e0f1e353dd)